compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -360,13 +391,37 @@ impl<'a> Parser<'a> {
             }
             Err(err) => return Err(err),
         };
-        let ty = if self.eat(&token::Semi) {
-            TyKind::Array(elt_ty, self.parse_anon_const_expr()?)
-        } else {
-            TyKind::Slice(elt_ty)
-        };
-        self.expect(&token::CloseDelim(token::Bracket))?;
-        Ok(ty)
+        let size_const =
+            if self.eat(&token::Semi) { Some(self.parse_anon_const_expr()?) } else { None };
+
+        match self.expect(&token::CloseDelim(token::Bracket)) {
+            Ok(_) => Ok(match size_const {
+                Some(size_const) => TyKind::Array(elt_ty, size_const),
+                None => TyKind::Slice(elt_ty),
+            }),
+            Err(mut err1)
+                if recover_array_type == RecoverArrayType::Yes
+                    && self.look_ahead(0, |t| t.kind == token::Comma)
+                    && size_const.is_none() =>
+            {
+                // Possibly used ',' instead of ';' in slice type syntax, try to recover
+                self.bump();
+                let snapshot = self.clone();
+                match self.parse_anon_const_expr() {
+                    Ok(array_size) => {
+                        err1.emit();
+                        self.expect(&token::CloseDelim(token::Bracket))?;
+                        Ok(TyKind::Array(elt_ty, array_size))
+                    }
+                    Err(mut err2) => {
+                        err2.cancel();
+                        *self = snapshot;
+                        Err(err1)
+                    }
+                }
```

Comment:
This... is unfortunate. I understand why you're using it. I'm the one that introduced this pattern into the compiler in the first place but it _should_ be used sparingly. I guess that ship might have sailed, and in that case we should then make this pattern less intrusive. :-/

Reply:
Which part do you mean? Snapshotting? Or recovery in general?